provider/openstack/firewaller.go

```diff
@@ -108,100 +108,130 @@ func (c *defaultFirewaller) SetUpGroups(controllerUUID, machineId string, apiPor
+				PortRangeMin: 1,
+				PortRangeMax: 65535,
 			},
 			{
+				Direction:  "ingress",
 				IPProtocol: "icmp",
-				FromPort:   -1,
-				ToPort:     -1,
 			},
 		})
 }
 
 // zeroGroup holds the zero security group.
-var zeroGroup nova.SecurityGroup
+var zeroGroup neutron.SecurityGroupV2
 
 // ensureGroup returns the security group with name and perms.
 // If a group with name does not exist, one will be created.
 // If it exists, its permissions are set to perms.
-func (c *defaultFirewaller) ensureGroup(name string, rules []nova.RuleInfo) (nova.SecurityGroup, error) {
-	novaClient := c.environ.nova()
+func (c *defaultFirewaller) ensureGroup(name string, rules []neutron.RuleInfoV2) (neutron.SecurityGroupV2, error) {
+	neutronClient := c.environ.neutron()
 	// First attempt to look up an existing group by name.
-	group, err := novaClient.SecurityGroupByName(name)
+	groupsFound, err := neutronClient.SecurityGroupByNameV2(name)
 	if err == nil {
-		// Group exists, so assume it is correctly set up and return it.
-		// TODO(jam): 2013-09-18 http://pad.lv/121795
-		// We really should verify the group is set up correctly,
-		// because deleting and re-creating environments can get us bad
-		// groups (especially if they were set up under Python)
-		return *group, nil
+		for _, group := range groupsFound {
+			if c.verifyGroupRules(group.Rules, rules) {
+				return group, nil
+			}
+		}
 	}
 	// Doesn't exist, so try and create it.
-	group, err = novaClient.CreateSecurityGroup(name, "juju group")
+	newGroup, err := neutronClient.CreateSecurityGroupV2(name, "juju group")
 	if err != nil {
-		if !gooseerrors.IsDuplicateValue(err) {
+		return zeroGroup, err
+	}
+	// The new group is created so now add the rules.
+	for _, rule := range rules {
+		rule.ParentGroupId = newGroup.Id
+		groupRule, err := neutronClient.CreateSecurityGroupRuleV2(rule)
+		if err != nil {
 			return zeroGroup, err
-		} else {
-			// We just tried to create a duplicate group, so load the existing group.
-			group, err = novaClient.SecurityGroupByName(name)
-			if err != nil {
-				return zeroGroup, err
-			}
-			return *group, nil
 		}
+		newGroup.Rules = append(newGroup.Rules, *groupRule)
 	}
-	// The new group is created so now add the rules.
-	group.Rules = make([]nova.SecurityGroupRule, len(rules))
-	for i, rule := range rules {
-		rule.ParentGroupId = group.Id
-		if rule.Cidr == "" {
-			// http://pad.lv/1226996 Rules that don't have a CIDR
-			// are meant to apply only to this group. If you don't
-			// supply CIDR or GroupId then openstack assumes you
-			// mean CIDR=0.0.0.0/0
-			rule.GroupId = &group.Id
+	return *newGroup, nil
+}
+
+// verifyGroupRules verifies the group rules against the rules we're looking for.
+func (c *defaultFirewaller) verifyGroupRules(rules []neutron.SecurityGroupRuleV2, rulesToMatch []neutron.RuleInfoV2) bool {
+	if len(rules) - len(rulesToMatch) != 2 {
+		return false
+	}
+	count := len(rulesToMatch)
+	for _, rule := range rules {
```

Comment:
Quadratic behaviour here, but we should have a pretty bounded number of rulesToMatch.